### PR TITLE
MK6 Charge System

### PR DESCRIPTION
**Pull request by Grey-Soldierman**
**When merged this pull request will:**
- Add a charge system for 82mm rounds used with the MK6 Mortar
- Preparation of ammunition:
    - Selecting the appropriate charge for a particular round. (Self interact menu)
    - Increasing charge 1 step requires 'ACE_Propelling_Charge' item
    - Decreasing charge 1 step gives 'ACE_Propelling_Charge' item
- Removal of charge firemodes on weapon 
     - Charges are instead handled at the magazine level
- Load Magazine actions
     - Switched to a generic function to generate the actions for loading compatible magazines. 
     - Lowered the time taken to load rounds when the charge system is enabled; the time is instead spent 
        on preparing the ammunition 
- Configurable 
     - Can be configured almost entirely through config values on the magazines & weapon. Number of 
      charges, number of charges required on each step, power of each charge. 
     - Enabled through settings menu.  Use of charges can be done separately of 'Ammo Handling' 
       however the loading of charges would be difficult without it. 

- Why?
Gives a slightly more realistic mortar gunner experience. Gives more purpose to assistant gunners and requires more thought to the operation of a mortar. Built to be expandable; 3rd party mods could take advantage of it and provide more accurate models for the different charge levels for a particular round. 

Images:
- Self interaction: https://imgur.com/B2ZpK2Q.jpg
- Interacting with mortar: https://i.imgur.com/1p2e9Kd.jpg
- Charge table: https://i.imgur.com/HRdvzh1.jpg

### CODE REVIEW

File: addons/mk6mortar/stringtable.xml [hunk 1]
```diff
@@ -512,5 +545,20 @@
...
+            <English>For use with 82mm mortar ammunition</English>
+        </Key>
+        <Key ID="STR_ACE_Mk6Mortar_useChargeSystem_DisplayName">
+            <English>Use Charge System</English>
+        </Key>
+        <Key ID="STR_ACE_Mk6Mortar_useChargeSystem_Description">
+            <English>Requires 82mm mortar shells to have their charges altered to achieve the desired range. Will affect AI</English>
+        </Key>
+        <Key ID="STR_ACE_Mk6Mortar_ammunition_action">
+            <English>Prepare Ammunition</English>
```
Comment (dedmen) [inline addons/mk6mortar/stringtable.xml]: Different from https://github.com/acemod/ACE3/pull/6577/files#diff-828b9969fd2828d749a5b198f91748baR278 ?

**Comment by Grey-Soldierman** [comment]
Cheers the comments Dedmen; It's what i love about ACE. Ill get them sorted today.

File: addons/mk6mortar/data/propelling_charge.rvmat [hunk 2]
```diff
@@ -77,3 +73,7 @@ class Stage7
 	texture="a3\data_f\env_land_co.paa";
 	uvSource="none";
 };
+class StageTI
+{
+	texture="a3\data_f\default_ti_ca.paa";
+};
```
Comment (bux) [inline addons/mk6mortar/data/propelling_charge.rvmat]: lacks blank line at EOF

File: addons/mk6mortar/stringtable.xml (hunk 1 above)
Comment (Grey-Soldierman) [reply addons/mk6mortar/stringtable.xml]: Im not following; preparing ammunition and to prepare ammunition is different

File: addons/mk6mortar/functions/fnc_rangeTablePreCalculatedValues.sqf [hunk 3]
```diff
@@ -20,19 +20,46 @@
 params ["_muzzleVelocity", "_airFriction"];
 
 switch (true) do {
-
+case ((abs(_muzzleVelocity - 50) < 0.00001) && {(abs(_airFriction - -0.0001) < 0.00001)}): {
```
[the remark this answers is not in the document]
Comment (Grey-Soldierman) [reply addons/mk6mortar/functions/fnc_rangeTablePreCalculatedValues.sqf]: Thats how they are generated by fnc_devBuildTable.

File: addons/mk6mortar/stringtable.xml [hunk 4]
```diff
@@ -273,6 +273,9 @@
             <Chinese>卸除彈頭中</Chinese>
         </Key>
         <Key ID="STR_ACE_Mk6Mortar_loadingMortar">
+            <English>Loading Round</English>
```
[the remark this answers is not in the document]
Comment (bux) [reply addons/mk6mortar/stringtable.xml]: eher "Lade Granate"

File: addons/mk6mortar/functions/fnc_prepRound.sqf [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+ * Example:
+ * [_player,'ACE_1Rnd_82mm_Mo_HE','ACE_1Rnd_82mm_Mo_HE_Charge_3','ACE_Propelling_Charge',1] call ace_mk6mortar_fnc_prepRound
+ *
+ * Public: Yes
+ */
+#include "script_component.hpp"
+
+params ["_unit","_oldMagazine","_newMagazine","_chargeClass","_chargesRequired"];
+
+if !(isNil (_newMagazine)) exitWith {ERROR("New magazine classname required");};
```
[the remark this answers is not in the document]
Comment (dedmen) [reply addons/mk6mortar/functions/fnc_prepRound.sqf]: Same classname thing in https://github.com/acemod/ACE3/pull/6577/files/81b3c317d04c54704dfeb52f22a5fda1b646de47..004ecab2eb7ff921aa45b358380282bb45aeeb81#diff-e61cd06a2f0e001acbc91e8de253c6cbR31
Are you really intending to check the classname string with isNil?

File: addons/mk6mortar/functions/fnc_prepRound.sqf [hunk 6]
```diff
@@ -20,43 +20,46 @@
...
 
-    if (_chargeCount < _chargesRequired) exitWith {ERROR("Player does not the required amount of charges")};
+    if (_chargeCount < _chargesRequired) exitWith {_canAddMagazine=false};
 
     for "_i" from 1 to _chargesRequired do {
         _unit removeItem _chargeClass;
     };
 };
 
+if !(_canAddMagazine) exitWith{ERROR("Player does not the required amount of charges")};
```
Comment (dedmen) [inline addons/mk6mortar/functions/fnc_prepRound.sqf]: whitespace after `exitWith`

File: addons/mk6mortar/functions/fnc_rangeTableOpen.sqf [hunk 7]
```diff
@@ -28,22 +28,23 @@ if (GVAR(useChargeSystem)) then {
...
     // Find all the base magazines with charges, store the velocity of the base magazine and each of their child charges
     {
-        private _baseCharge  = getNumber(configFile >> "CfgMagazines" >> _x >> QGVAR(baseCharge));
+        private _currentMagazine = _cfgMagazines >> _x;
+        private _baseCharge  = getNumber(_currentMagazine QGVAR(baseCharge));
         if (_baseCharge == 1) then {
-            private _childCharges = getArray(configFile >> "CfgMagazines" >> _x >> QGVAR(charges));
-            private _magName = getText (configFile >> "CfgMagazines" >> _x >> "displayNameShort");
-            private _initSpeed = getNumber (configFile >> "CfgMagazines" >> _x >> "initSpeed");
+            private _childCharges = getArray(_currentMagazine QGVAR(charges));
```
Comment (dedmen) [inline addons/mk6mortar/functions/fnc_rangeTableOpen.sqf]: missing `>>`

Comment (dedmen) [reply addons/mk6mortar/functions/fnc_rangeTableOpen.sqf]: Also below and above in L35

File: addons/mk6mortar/functions/fnc_prepRound.sqf (hunk 5 above)
[the remark this answers is not in the document]
Comment (Grey-Soldierman) [reply addons/mk6mortar/functions/fnc_prepRound.sqf]: The intention was to make sure there is a classname to replace the old the magazine with, so if people setup the config incorrectly, add magazine won't complain that its not got nothing to add

File: addons/mk6mortar/functions/fnc_prepRound.sqf [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+            _unit addItem _chargeClass;
+        };
+    } else {
+        _pos = _unit modelToWorldVisual [0.5,0.5,0]; // Front right of player
+        _holder = createVehicle ["WeaponHolder_Single_F",_pos,[],0,"NONE"];
+        _holder addItemCargo [_chargeClass, _chargesRequired];
+        _holder setPosATL _pos;
+    };
+} else {
+    private _chargeCount = {_x == _chargeClass} count items _unit;
```
[the remark this answers is not in the document]
Comment (Grey-Soldierman) [reply addons/mk6mortar/functions/fnc_prepRound.sqf]: Not sure of an alternative really, I haven't see another way to check whether a player has X amount of a given item.  The only alternative i can think of is to switch the way it does the magazine swap to a loop;

Each loop will take the given magazine 1 iteration in the number of charges, each loop would check the player has the propelling charge item, remove it if the item is present or stop the loop if the player doesn't have the required charge item. Presumably the progress bar system could be used to achieve this

File: addons/mk6mortar/functions/fnc_rangeTablePreCalculatedValues.sqf (hunk 3 above)
[the remark this answers is not in the document]
Comment (dedmen) [reply addons/mk6mortar/functions/fnc_rangeTablePreCalculatedValues.sqf]: Someone should maybe update that thing then :D

File: addons/mk6mortar/functions/fnc_prepRound.sqf [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    
+    //Flip the charges required from negative to positive number so we can easily check against it
+    _chargesRequired = - _chargesRequired;
+
+    if (_chargeCount < _chargesRequired) exitWith{_canAddMagazine=false};
+
+    for "_i" from 1 to _chargesRequired do {
```
Comment (dedmen) [inline addons/mk6mortar/functions/fnc_prepRound.sqf]: Bug. Removing items even though they are never used if player doesn't have enough of them.

File: addons/mk6mortar/functions/fnc_prepRound.sqf (hunk 8 above)
[the remark this answers is not in the document]
Comment (dedmen) [reply addons/mk6mortar/functions/fnc_prepRound.sqf]: It's probably not a big problem here anyway.
Could use https://community.bistudio.com/wiki/getItemCargo for all the containers (uniform,vest,backpack) to get item counts. It's alot faster than iterating through `items`.
But that's quite alot of code and probably not really worth it.

File: addons/mk6mortar/stringtable.xml (hunk 1 above)
Comment (dedmen) [reply addons/mk6mortar/stringtable.xml]: Yes that was the question. Fine then.

File: addons/mk6mortar/functions/fnc_prepRound.sqf (hunk 9 above)
Comment (dedmen) [reply addons/mk6mortar/functions/fnc_prepRound.sqf]: Nvm. `exitWith`.

File: addons/mk6mortar/functions/fnc_prepRound.sqf [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        _unit removeItem _chargeClass;
+    };
+};
+
+if !(_canAddMagazine) exitWith{ERROR("Player does not the required amount of charges")};
```
Comment (VKing6) [inline addons/mk6mortar/functions/fnc_prepRound.sqf]: Does not _have_?

**Comment by Grey-Soldierman** [comment]
The main reason I left the charge 4 magazines as they are is to not confuse players when the charge system is deactivated. Ill have think and see if i can append something within the actions to make it clearer.

**Comment by severgun** [comment]
If I understand correctly, main point of this system is realism.
Then...
1) Would be nice to get rid of player inventory mess. IMHO rounds should be prepaired from boxes or ground. What if I have mortar, ammobox and no backpack? ACE2 system was great at this.(there also were different detonators)
2) Would be nice to see WIKI page about creating charges for third-party mortars. (If it is possible. If not - then turn system into framework)
3) There is advanced throwing, attach, refuel, fortify, etc where player can carry object in hands. Isn't it better than https://i.imgur.com/1p2e9Kd.jpg ?

**Comment by PabstMirror** [comment]
Made some changes here: https://github.com/acemod/ACE3/commit/158fbe49f6969f762f2a45b6982a38c75e10f0d9
CBA Settings, 
hide's mortar hud charge display, 
add condition checking for showing prep actions
other minor stuff
let me know if these look ok

As for the mag names, it is a little confusing knowing that the default round is full charge, but I also see the problem in changing the name and confusing the other mode.
Maybe we could at least make a change in the interaction menu, if we don't want to change the configs?

I also think the lowest charge should be "0" instead of "1" (it has no charges attached, but still has the internal propellant)
ref: https://fas.org/man/dod-101/sys/land/m821.htm

**Comment by jonpas** [comment]
Closing due to age and/or inactivity, unlikely to be finished or has to be reworked due to other changes over the years. May be re-opened if someone finds it useful, even better just open a new up-to-date PR.

**Pull request closed without merging**